portfolio/src/main/java/com/google/sps/servlets/HomeServlet.java

```diff
@@ -0,0 +1,85 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+
+@WebServlet("/home")
+public class HomeServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    UserService userService = UserServiceFactory.getUserService();
+    
+
+    // If user is not logged in, show a login form (could also redirect to a login page)
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL("/home"); 
+      out.println("<p>You must <a href=\"" + loginUrl + "\">login</a> to comment</p>");
+    //   response.sendRedirect("/index.html");
+      return;
+    }
+ 
+    // If user has not set a nickname, redirect to nickname page
+    String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+    if (nickname == null) {
+      response.sendRedirect("/nickname"); 
+      //in implementation, redirect to index
+      return;
+    }
+
+    // User is logged in and has a nickname, so the request can proceed
+    String logoutUrl = userService.createLogoutURL("/home"); 
+    List<String> login = new ArrayList<>();
+    login.add(nickname);
+    login.add(logoutUrl);
+    
+    // out.println("<p>Hello " + nickname + "!</p>");
+    // out.println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+    // out.println("<p>Change your nickname <a href=\"/nickname\">here</a>.</p>");
+
+     response.setContentType("application/json;");
+     response.getWriter().println(login);
+     response.sendRedirect("/index.html");
+  }
+
+  /** Returns the nickname of the user with id, or null if the user has not set a nickname. */
+  private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return null;
+    }
+    String nickname = (String) entity.getProperty("nickname");
+    return nickname;
+  }
```

Comment:
You could use a try and catch block here, instead of checking for entity null and catch on NullPointerException.